src/backend/linux_raw/fs/raw_dir.rs

```diff
@@ -0,0 +1,204 @@
+use core::cmp::max;
+use core::convert::TryFrom;
+use core::mem::MaybeUninit;
+use core::slice;
+use linux_raw_sys::general::linux_dirent64;
+
+use crate::backend::fs::syscalls::getdents_uninit;
+use crate::fd::AsFd;
+use crate::ffi::CStr;
+use crate::fs::FileType;
+use crate::io;
+
+/// Suggested buffer size for use in [`RawDir::new`]. There is no requirement to
+/// use this buffer size, but it was found to be a good tradeoff between memory
+/// use and syscall count.
+pub const DIR_BUF_LEN: usize = 8192;
+
+/// A directory iterator implemented with getdents.
+///
+/// This implementation:
+/// - Excludes deleted inodes (with ID 0).
+/// - Does not handle growing the buffer. If this functionality is necessary,
+///   you'll need to drop the current iterator, resize the buffer, and then
+///   re-create the iterator. The iterator is guaranteed to continue where it
+///   left off provided the file descriptor isn't changed. See the example in
+///   [`RawDir::new`].
+pub struct RawDir<'buf, Fd: AsFd> {
+    fd: Fd,
+    buf: &'buf mut [MaybeUninit<u8>],
+    initialized: usize,
+    offset: usize,
+}
+
+impl<'buf, Fd: AsFd> RawDir<'buf, Fd> {
+    /// Create a new iterator from the given file descriptor and buffer.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::mem::MaybeUninit;
+    /// # use rustix::fs::{cwd, Mode, OFlags, openat, RawDir};
+    ///
+    /// let fd = openat(cwd(), ".", OFlags::RDONLY | OFlags::DIRECTORY, Mode::empty()).unwrap();
+    ///
+    /// let mut buf = [MaybeUninit::uninit(); 2048];
```

Comment:
Can this example use `DIR_BUF_LEN`?

Reply:
Before making docs changes, I want to make sure we're on the same page since I think we're coming at this from different angles. I was trying to demo the edge cases while assuming that people would know to use `Vec::spare_capacity_mut` for everyday life, but that's probably a bad assumption.

Here's how I'd like people to use the API:
- When using recursion, people should either use a vec or the stack. If using the stack, I'd lean towards a smaller value to minimize wasted space and skip [stack probes](https://docs.rs/compiler_builtins/latest/compiler_builtins/probestack/index.html), hence the 2048. If using the heap, then we can afford to waste more space, hence using 8192. So maybe this suggests having a constant is actually a bad idea? Instead we could tell people to "Use a buffer size of at least NAME_MAX+24 bytes. We suggest 2048 for stack allocated buffers and 8192 for heap allocated buffers." In practice, almost all file systems use 255 or smaller as their limit: https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits. Reiser appears to be the only deviant.
- When not using recursion, people would ideally have a cached vec somewhere that they re-use.

> Can we guarantee that DIR_BUF_LEN is long enough to support any path that the host OS supports (considering NAME_MAX)?

I don't think so unfortunately, which is why I'm leaning towards removing the constant. There are basically no guarantees on NAME_MAX: https://www.gnu.org/software/libc/manual/html_node/Limits-for-Files.html. Wikipedia has the current file system limits, but nothing prevents a future file system from removing the limit entirely for example.

Going back to the docs, what about having a simple heap example, a simple stack example, and then a production-ready buffer resizing example?

Reply:
> Going back to the docs, what about having a simple heap example, a simple stack example, and then a production-ready buffer resizing example?

That sounds good. I imagine you could even skip the simple heap example if you wanted to. I imagine the vast majority of Rust code will continue using `std::fs::ReadDir` to read directories, or [walkdir](https://crates.io/crates/walkdir) or so, so I imagine the main audience for using this API directly will be people doing low-level optimization work.

For the stack example, I'm not comfortable encouraging people to use fixed-sized buffers if OS's don't have a limit. If we're not confident enough to assume a `NAME_MAX` exists, it feels awkward to make suggestions to users that they bake in numbers like 2048. Could we instead give some guidance like, "only use this approach for reading directories with known layouts where all entries have names less than X" or so?

Reply:
> Could we instead give some guidance like, "only use this approach for reading directories with known layouts where all entries have names less than X" or so?

I think that's reasonable, I added clarifications for the simple stack + heap examples. Happy to remove either if you'd rather not have the example at all.